### ISSUE TITLE
toString method is no longer present (or visible), causing runtime exceptions

### ISSUE THREAD

**Issue by jimshowalter** [square/okhttp#5426]
We upgraded to the latest, and now our build fails with:

[2019-09-06T01:15:49.953Z] Caused by: java.lang.NoSuchMethodError: okhttp3.Cookie.toString(Z)Ljava/lang/String;
[2019-09-06T01:15:49.953Z] at okhttp3.JavaNetCookieJar.saveFromResponse(JavaNetCookieJar.java:43)
[2019-09-06T01:15:49.953Z] at okhttp3.internal.http.HttpHeaders.receiveHeaders(HttpHeaders.kt:207)
[2019-09-06T01:15:49.953Z] at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:86)
[2019-09-06T01:15:49.953Z] at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:112)
[2019-09-06T01:15:49.953Z] at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:71)
[2019-09-06T01:15:49.953Z] at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:112)
[2019-09-06T01:15:49.953Z] at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:87)
[2019-09-06T01:15:49.953Z] at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.kt:184)
[2019-09-06T01:15:49.953Z] at okhttp3.RealCall.execute(RealCall.kt:66)
[2019-09-06T01:15:49.953Z] at com.blackducksoftware.integration.hub.rest.CredentialsRestConnection.clientAuthenticate(CredentialsRestConnection.java:86)
[2019-09-06T01:15:49.953Z] at com.blackducksoftware.integration.hub.rest.RestConnection.connect(RestConnection.java:128)
[2019-09-06T01:15:49.953Z] at com.blackducksoftware.integration.hub.rest.RestConnection.handleExecuteClientCall(RestConnection.java:350)
[2019-09-06T01:15:49.953Z] at com.blackducksoftware.integration.hub.rest.RestConnection.handleExecuteClientCall(RestConnection.java:317)
[2019-09-06T01:15:49.953Z] at com.blackducksoftware.integration.hub.request.HubRequest.executePost(HubRequest.java:85)
[2019-09-06T01:15:49.953Z] at com.blackducksoftware.integration.hub.api.bom.BomImportRequestService.importBomFile(BomImportRequestService.java:55)
[2019-09-06T01:15:49.953Z] at com.blackducksoftware.integration.hub.buildtool.BuildToolHelper.deployHubOutput(BuildToolHelper.java:71)
[2019-09-06T01:15:49.953Z] at com.blackducksoftware.integration.gradle.task.BuildBomTask.deployHubBDIO(BuildBomTask.java:240)
[2019-09-06T01:15:49.953Z] at com.blackducksoftware.integration.gradle.task.BuildBomTask.performTask(BuildBomTask.java:153)
[2019-09-06T01:15:49.953Z] at com.blackducksoftware.integration.gradle.task.BuildBomTask.task(BuildBomTask.java:130)
[2019-09-06T01:15:49.953Z] at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:103)

We don't see in the release notes anything saying that method has been removed.

Can you put it back? If not, we can't upgrade to your latest, because we need to run blackduck until we replace it (long process).

You replied "Presumably you are mixing a 3.14 version of JavaNetCookieJar from okhttp-httpurlconnection with 4.x of okhttp. Can you upgrade both?" and then asked for more info and then closed it 8 minutes later. I was in standup. And I can't reopen a bug closed by a code owner. So here it is again.

./gradlew dependencies shows no occurrence of okhttp-httpurlconnection on any of our classpaths.

We are only using (or trying to use) com.squareup.okhttp3:okhttp:4.1.1.

**Comment by yschimke** [comment]
Sorry, it's just okhttp-urlconnection.  But you can't mix these versions.  That is where okhttp3.JavaNetCookieJar comes from.

I closed the previous one as I had high confidence that is the problem, and it's not a bug in okhttp.  Just mixing versions.

**Comment by jimshowalter** [comment]
I understand about classpath conflicts, diamond problems, class shadowing, jar hell. When you say we can't mix versions, I totally agree, but don't know versions of what. We only have okhttp3 on the classpath. There's no other okhttp jar. I can dump the classpath and grep -E whatever pattern you want to see what's on it that might be conflicting.

**Comment by yschimke** [comment]
Can you use your IDE to see which jar is providing it?

<img width="671" alt="image" src="https://user-images.githubusercontent.com/231923/64460535-81cefd80-d0f2-11e9-960f-a0be281a9e14.png">

See above it is not in the core okhttp jar.

**Comment by jimshowalter** [comment]
It's not present anywhere in the classpath in my IDE (which is IntelliJ). To look for it just requires CMD o, and the class name. When I do that for Cookie and search down in the list, I get the com.squareup.okhttp3 jar. When I do that for JavaNetCookieJar, there's no match. It's also not anywhere in the Gradle 5.5.1 uber-jar.

**Comment by yschimke** [comment]
It's showing up in your stacktrace

```
[2019-09-06T01:15:49.953Z] Caused by: java.lang.NoSuchMethodError: okhttp3.Cookie.toString(Z)Ljava/lang/String;
[2019-09-06T01:15:49.953Z] at okhttp3.JavaNetCookieJar.saveFromResponse(JavaNetCookieJar.java:43)
[2019-09-06T01:15:49.953Z] at okhttp3.internal.http.HttpHeaders.receiveHeaders(HttpHeaders.kt:207)
```

**Comment by jimshowalter** [comment]
Yes it is. I'm just saying that it's coming from some unknown location. Anyway, I found the fix. We had added a closure of jars to the buildscript's dependencies section, and that dragged in the above class from somewhere. Reverting that fixed the problem (but without explaining where JavaNetCookieJar came from). I even dumped the classpath and it wasn't there.

**Comment by nasajoey** [comment]
Not suggesting to reopen, but I had different fix in case other stumble on this like I did.  I started with an open source project that needed okhttp3 and retrofit2.  Got the same exception on first building.  I had to use the version of okhttp3 called out as a dependency for most recent retrofit2:

https://mvnrepository.com/artifact/com.squareup.retrofit2/retrofit/2.6.2

it says okhttp3 3.12.0 is a dependency, so I changed my pom to point to that version rather than the latest and the exception went away.  I am assuming that is what the "mixing versions" discussion above is about, but not clear that these two separate projects would stomp on each other in this way... probably a bug report in this for someone somewhere.
